x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/SimplifyComparisonsArithmetics.java

```diff
@@ -0,0 +1,244 @@
+        if (expression.foldable()) {
+            try {
+                expression = new Literal(expression.source(), expression.fold(), expression.dataType());
+            } catch (ArithmeticException | DateTimeException e) {
+                // null signals that folding would result in an over-/underflow (such as Long.MAX_VALUE+1); the optimisation is skipped.
+                expression = null;
+            }
+        }
+        return expression;
+    }
+
+    private abstract static class OperationSimplifier {
+        final BinaryComparison comparison;
+        final Literal bcLiteral;
+        final ArithmeticOperation operation;
+        final Expression opLeft;
+        final Expression opRight;
+        final Literal opLiteral;
+
+        OperationSimplifier(BinaryComparison comparison) {
+            this.comparison = comparison;
+            operation = (ArithmeticOperation) comparison.left();
+            bcLiteral = (Literal) comparison.right();
+
+            opLeft = operation.left();
+            opRight = operation.right();
+
+            if (opLeft instanceof Literal) {
+                opLiteral = (Literal) opLeft;
+            } else if (opRight instanceof Literal) {
+                opLiteral = (Literal) opRight;
+            } else {
+                opLiteral = null;
+            }
+        }
+
+        // can it be quickly fast-tracked that the operation can't be reduced?
+        final boolean isUnsafe(BiFunction<DataType, DataType, Boolean> typesCompatible) {
+            if (opLiteral == null) {
+                // one of the arithm. operands must be a literal, otherwise the operation wouldn't simplify anything
+                return true;
+            }
+
+            // Only operations on fixed point literals are supported, since optimizing float point operations can also change the
+            // outcome of the filtering:
+            // x + 1e18 > 1e18::long will yield different results with a field value in [-2^6, 2^6], optimised vs original;
+            // x * (1 + 1e-15d) > 1 : same with a field value of (1 - 1e-15d)
+            // so consequently, int fields optimisation requiring FP arithmetic isn't possible either: (x - 1e-15) * (1 + 1e-15) > 1.
+            if (opLiteral.dataType().isRational() || bcLiteral.dataType().isRational()) {
+                return true;
+            }
+
+            // the Literal will be moved to the right of the comparison, but only if data-compatible with what's there
+            if (typesCompatible.apply(bcLiteral.dataType(), opLiteral.dataType()) == false) {
+                return true;
+            }
+
+            return isOpUnsafe();
+        }
+
+        final Expression apply() {
+            // force float point folding for FlP field
+            Literal bcl = operation.dataType().isRational()
+                ? new Literal(bcLiteral.source(), ((Number) bcLiteral.value()).doubleValue(), DataTypes.DOUBLE)
```

Comment:
Changing this to use `new Literal` instead of `Literal.of` resolves https://github.com/elastic/elasticsearch/issues/108388